04/number-of-provinces.md

```diff
@@ -0,0 +1,223 @@
+##### !question
+
+1. Use the comments provided to write at least two example input/outputs:
+    * Consider at least one nominal and one edge case.
+    * What is the expected output for the given input?
+    * You can use the examples provided in the prompt, or other examples.
+2. Write unit tests for `num_provinces` for the nominal and edge cases you identified in the first step.
+
+*Note: Click the **Run Tests** button to save your tests for instructor feedback. No real tests are actually run again your unit tests.*
+
+##### !end-question
+##### !placeholder
+
+```py
+# example input 1:
+# expected output 1:
+
+# example input 2:
+# expected output 2:
+
+def test_nominal_case():
+    # ^rename with meaningful test name
+    # and complete test implementation below
+    pass
+    # arrange
+
+    # act
+
+    # assert
+
+def test_edge_case():
+    # ^rename with meaningful test name
+    # and complete test implementation below
+    pass
+    
+    # arrange
+    
+    # act
+    
+    # assert
+```
+##### !end-placeholder
+
+##### !tests
+
+```py
+import unittest
+
+class TestPython1(unittest.TestCase):
+  def test_always_pass(self):
+    self.assertEqual(1,1)
+```
+
+##### !end-tests
+##### !explanation 
+
+Example tests:
+
+```python
+def test_returns_two_for_two_provinces():
+    # Arrange
+    is_connected = [[0, 0, 1], [0, 0, 0], [1, 0, 0]]
+
+    # Act
+    result = num_provinces(is_connected)
+
+    # Assert
+    assert result == 2
+
+def test_returns_zero_for_empty_connections():
+    # Arrange
+    is_connected = []
+
+    # Act
+    result = num_provinces(is_connected)
+
+    # Assert
+    assert result == 0
+```
+
+##### !end-explanation
+##### !hint
+Try drawing out some sample graphs and translating them into an adjacency matrix if coming up with an adjacency matrix representing a graph is challenging.
+##### !end-hint
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 3 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: 647614a1-13b7-42f8-aaa4-4487184cac43
+* title: Create Logical Steps
+* topics: pse
+##### !question
+
+Without writing code, describe how you would implement `num_provinces` in enough detail that someone else could write the code. 
+* It may be helpful to break up the problem/algorithm into smaller subproblems/algorithms. For example, 1. Handle invalid input, 2. Given valid input, perform the computation/solve the problem/etc.
+* Your logical steps could take the form of a numbered list, pseudo code, or anywhere in between. What's important at this stage is to think through and outline the implementation before writing code.
+
+##### !end-question
+
+##### !placeholder
+
+Write the logical steps here.
+
+##### !end-placeholder
+
+##### !hint
+When thinking about whether two cities are in the same province, do we need to consider all pairwise cities, or is there another way for us to find all of the cities in a particular province?
+##### !end-hint
+
+##### !hint
```

Comment:
I really like the gradual hint usage, I think this could be quite helpful.

Reply:
Me too! For self-guided, asynchronous work it feels really handy for gradually building to a solution. I'd like to see us do this more in the main curriculum, though I do get concerned that could become a crutch for folks to avoid asking questions.